test/e2e/volumes.go

```diff
@@ -0,0 +1,360 @@
+		}
+
+		defer func() {
+			if clean {
+				volumeTestCleanup(c, config)
+			}
+		}()
+
+		pod := startVolumeServer(c, config)
+		if pod == nil {
+			// Skip the test, message has been already logged.
+			return
+		}
+
+		serverIP := pod.Status.PodIP
+		Logf("NFS server IP address: %v", serverIP)
+
+		volume := api.VolumeSource{
+			NFS: &api.NFSVolumeSource{
+				Server:   serverIP,
+				Path:     "/",
+				ReadOnly: true,
+			},
+		}
+		testVolumeClient(c, config, volume, "Hello from NFS!")
+	})
+})
+
+////////////////////////////////////////////////////////////////////////////////
+// Gluster
+////////////////////////////////////////////////////////////////////////////////
+var _ = Describe("volumes-Glusterfs", func() {
+	clean := true // If 'false', the test won't clear its namespace (and pods and services) upon completion. Useful for debugging.
+
+	// filled in BeforeEach
+	var c *client.Client
+	var namespace *api.Namespace
+
+	BeforeEach(func() {
+		var err error
+		c, err = loadClient()
+		Expect(err).NotTo(HaveOccurred())
+		By("Building a namespace api object")
+		namespace, err = createTestingNS("volume-gluster", c)
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	AfterEach(func() {
+		if err := c.Namespaces().Delete(namespace.Name); err != nil {
+			Failf("Couldn't delete ns %s", err)
+		}
+	})
+
+	It("should be able to mount GlusterFS", func() {
+		config := VolumeTestConfig{
+			namespace:   namespace.Name,
+			prefix:      "gluster",
+			serverImage: "gcr.io/google_containers/volume-gluster",
+			serverPorts: []int{24007, 24008, 49152},
+		}
+
+		defer func() {
+			if clean {
+				volumeTestCleanup(c, config)
+			}
+		}()
+
+		pod := startVolumeServer(c, config)
+		if pod == nil {
+			// Skip the test, message has been already logged.
+			return
+		}
+
+		serverIP := pod.Status.PodIP
+		Logf("Gluster server IP address: %v", serverIP)
+
+		// create Endpoints for the server
+		endpoints := api.Endpoints{
+			TypeMeta: api.TypeMeta{
+				Kind:       "Endpoints",
+				APIVersion: "v1beta3",
+			},
+			ObjectMeta: api.ObjectMeta{
+				Name: config.prefix + "-server",
+			},
+			Subsets: []api.EndpointSubset{
+				{
+					Addresses: []api.EndpointAddress{
+						{
+							IP: serverIP,
+						},
+					},
+					Ports: []api.EndpointPort{
+						{
+							Name:     "gluster",
+							Port:     24007,
+							Protocol: api.ProtocolTCP,
+						},
+					},
+				},
+			},
+		}
+
+		endClient := c.Endpoints(config.namespace)
+
+		defer func() {
+			if clean {
+				endClient.Delete(config.prefix + "-server")
+			}
+		}()
+
+		if _, err := endClient.Create(&endpoints); err != nil {
+			Failf("Failed to create endpoints for Gluster server: %v", err)
+		}
+
+		volume := api.VolumeSource{
+			Glusterfs: &api.GlusterfsVolumeSource{
+				// Use service name as an endpoint
+				EndpointsName: config.prefix + "-server",
+				Path:          "test_vol",
+				ReadOnly:      true,
+			},
+		}
+		testVolumeClient(c, config, volume, "Hello from Gluster!")
```

Comment:
Add a comment that this must be kept consistent with contrib/for-tests/volumes/gluster/index.html